File: plugins/sd-wiki/scripts/wiki.py
"""원격 위키 CLI (플러그인 sd-wiki).

에이전트가 Bash 로 능동 호출하는 진입점. 인자 파싱 → `wiki_core` 위임 → JSON 출력만
담당. 원격 호출·인증·충돌 재시도 등 메커니즘은 전부 `wiki_core` 에 있고 이 파일엔 없음
(명령 추가·변경 시 이 파일만 보면 됨).

  python "${CLAUDE_PLUGIN_ROOT}/scripts/wiki.py" <명령> ...

stdout 은 서비스 응답 JSON, 오류는 stderr + 비0 종료코드.
"""
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path
from typing import Any

# 에이전트의 일반 셸엔 CLAUDE_PLUGIN_* env 가 없으므로 코어는 env 가 아니라 형제 경로로
# 찾음 — 직접 실행 시 sys.path[0] 이 이미 scripts/ 지만, 다른 cwd·import 경로에서도
# 견고하도록 자신의 디렉터리를 보강한다.
_SCRIPTS_DIR = str(Path(__file__).resolve().parent)
if _SCRIPTS_DIR not in sys.path:
    sys.path.insert(0, _SCRIPTS_DIR)

import wiki_core  # noqa: E402


for _stream in (sys.stdout, sys.stderr):
    try:
        _stream.reconfigure(encoding="utf-8")
    except Exception:
        pass


def _print_json(data: Any) -> None:
    sys.stdout.write(json.dumps(data, ensure_ascii=False, indent=2))
    sys.stdout.write("\n")


def _read_body_arg(args: argparse.Namespace) -> str:
    if args.body is not None and args.body_file is not None:
        raise wiki_core.WikiApiError("--body 와 --body-file 은 함께 쓸 수 없습니다.")
    if args.body is not None:
        return args.body
    if args.body_file is not None:
        try:
            return Path(args.body_file).read_text(encoding="utf-8")
        except OSError as err:
            raise wiki_core.WikiApiError(f"본문 파일을 읽을 수 없습니다: {err}") from err
    if not sys.stdin.isatty():
        return sys.stdin.read()
    raise wiki_core.WikiApiError("본문은 --body, --body-file 또는 stdin 으로 입력해야 합니다.")


def _build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="원격 위키 CLI")
    parser.add_argument(
        "--no-browser",
        action="store_true",
        help="토큰이 없을 때 브라우저 로그인을 띄우지 않고 종료합니다.",
    )

    subparsers = parser.add_subparsers(dest="command", required=True)

    read_parser = subparsers.add_parser("read", help="topic 으로 페이지를 읽습니다.")
    read_parser.add_argument("topic")

    write_parser = subparsers.add_parser("write", help="페이지를 생성하거나 갱신합니다.")
    write_parser.add_argument("topic")
    write_parser.add_argument("--title", required=True)
    write_parser.add_argument("--summary", required=True)
    write_parser.add_argument("--body")
    write_parser.add_argument("--body-file")
    write_parser.add_argument("--base-version", type=int)
    write_parser.add_argument("--parent", help="상위 페이지 topic. 생략 시 기존 상위 유지(신규는 최상위).")

    search_parser = subparsers.add_parser("search", help="키워드로 페이지를 검색합니다.")
    search_parser.add_argument("keyword")

    subparsers.add_parser("toc", help="목차를 조회합니다.")
    subparsers.add_parser("rootmap", help="최상위 노드의 라우팅 목록을 조회합니다.")

    children_parser = subparsers.add_parser("children", help="직속 자식 노드의 라우팅 목록을 조회합니다.")
    children_parser.add_argument("topic")

    delete_parser = subparsers.add_parser("delete", help="페이지를 삭제합니다(자식은 상위로 재배치).")
    delete_parser.add_argument("topic")
    delete_parser.add_argument("--base-version", type=int)

    move_parser = subparsers.add_parser("move", help="내용 변경 없이 상위 페이지만 이동합니다(순수 이동).")
    move_parser.add_argument("topic")
    move_group = move_parser.add_mutually_exclusive_group(required=True)
    move_group.add_argument("--parent", help="새 상위 페이지 topic.")
    move_group.add_argument("--root", action="store_true", help="최상위(루트)로 이동.")

    subparsers.add_parser("lint", help="위키 무결성·트리·링크를 점검합니다.")
    return parser


def _run_command(args: argparse.Namespace, token: str) -> Any:
    if args.command == "read":
        return wiki_core.call_service("read", [args.topic], token)
    if args.command == "search":
        return wiki_core.call_service("search", [args.keyword], token)
    if args.command == "toc":
        return wiki_core.call_service("toc", [], token)
    if args.command == "rootmap":
        return wiki_core.call_service("rootMap", [], token)
    if args.command == "children":
        return wiki_core.call_service("children", [args.topic], token)
    if args.command == "write":
        input_data: dict[str, Any] = {
            "topic": args.topic,
            "title": args.title,
            "summary": args.summary,
            "body": _read_body_arg(args),
        }
        if args.base_version is not None:
            input_data["baseVersion"] = args.base_version
        if args.parent is not None:
            input_data["parentTopic"] = args.parent
        return wiki_core.write_page(input_data, token)
    if args.command == "delete":
        delete_input: dict[str, Any] = {"topic": args.topic}
        if args.base_version is not None:
            delete_input["baseVersion"] = args.base_version
        return wiki_core.call_service("delete", [delete_input], token)
    if args.command == "move":
        parent_topic = None if args.root else args.parent
        return wiki_core.call_service("move", [{"topic": args.topic, "parentTopic": parent_topic}], token)
    if args.command == "lint":
        return wiki_core.call_service("lint", [], token)
    raise wiki_core.WikiApiError(f"알 수 없는 명령: {args.command}")


def _main(argv: list[str]) -> int:
    parser = _build_parser()
    args = parser.parse_args(argv)
    allow_browser = not args.no_browser

    try:
        token = wiki_core.get_token(allow_browser=allow_browser)
        if token is None:
            return 1
        try:
            result = _run_command(args, token)
        except wiki_core.WikiAuthExpired:
            if not allow_browser:
                raise
            token = wiki_core.browser_login()
            result = _run_command(args, token)
    except wiki_core.WikiWriteConflict as err:
        # 충돌은 실패(비0 종료)로 알리되, 재통합에 필요한 최신 본문을 stdout 으로 함께 전달.
        _print_json({"conflict": True, "message": str(err), "latest": err.latest})
        return 3
    except wiki_core.WikiAuthError as err:
        print(f"위키 인증 오류: {err}", file=sys.stderr)
        return 2
    except wiki_core.WikiApiError as err:
        print(f"위키 API 오류: {err}", file=sys.stderr)
        return 2

    _print_json(result)
    return 0


if __name__ == "__main__":
    sys.exit(_main(sys.argv[1:]))

File: plugins/sd-wiki/scripts/wiki.ts
/** 원격 위키 CLI (플러그인 sd-wiki).
 *
 * 에이전트가 Bash 로 능동 호출하는 진입점. 인자 파싱 → `wiki_core` 위임 → JSON 출력만
 * 담당. 원격 호출·인증·충돌 재시도 등 메커니즘은 전부 `wiki_core` 에 있고 이 파일엔 없음
 * (명령 추가·변경 시 이 파일만 보면 됨).
 *
 *   bun "${CLAUDE_PLUGIN_ROOT}/scripts/wiki.ts" <명령> ...
 *
 * stdout 은 서비스 응답 JSON, 오류는 stderr + 비0 종료코드.
 */

import { readFile } from "node:fs/promises";
import {
  WikiApiError,
  WikiAuthError,
  WikiAuthExpired,
  WikiWriteConflict,
  browserLogin,
  callService,
  getToken,
  writePage,
} from "./wiki_core.ts";

interface BaseArgs {
  readonly command: string;
  readonly noBrowser: boolean;
}

interface TopicArgs extends BaseArgs {
  readonly topic: string;
}

interface SearchArgs extends BaseArgs {
  readonly keyword: string;
}

interface WriteArgs extends TopicArgs {
  readonly title: string;
  readonly summary: string;
  readonly body?: string;
  readonly bodyFile?: string;
  readonly baseVersion?: number;
  readonly parent?: string;
}

interface MoveArgs extends TopicArgs {
  readonly parent?: string;
  readonly root: boolean;
}

type CliArgs = BaseArgs | TopicArgs | SearchArgs | WriteArgs | MoveArgs;

class CliParseError extends Error {}

function printJson(data: unknown): void {
  process.stdout.write(JSON.stringify(data ?? null, undefined, 2));
  process.stdout.write("\n");
}

function decodeUtf8Strict(data: Buffer | Uint8Array | ArrayBuffer): string {
  return new TextDecoder("utf-8", { fatal: true }).decode(data);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isFileReadError(error: unknown): boolean {
  return isRecord(error) && typeof error["code"] === "string";
}

async function readBodyFile(filePath: string): Promise<string> {
  let bytes: Buffer;
  try {
    bytes = await readFile(filePath);
  } catch (error) {
    if (isFileReadError(error)) {
      throw new WikiApiError(`본문 파일을 읽을 수 없습니다: ${getErrorMessage(error)}`);
    }
    throw error;
  }
  return decodeUtf8Strict(bytes);
}

function getErrorMessage(error: unknown): string {
  if (error instanceof Error && error.message) return error.message;
  return String(error);
}

async function readStdin(): Promise<string> {
  return await new Promise((resolve, reject) => {
    let data = "";
    process.stdin.setEncoding("utf8");
    process.stdin.on("data", (chunk) => {
      data += String(chunk);
    });
    process.stdin.on("end", () => resolve(data));
    process.stdin.on("error", reject);
  });
}

async function readBodyArg(args: WriteArgs): Promise<string> {
  if (args.body !== undefined && args.bodyFile !== undefined) {
    throw new WikiApiError("--body 와 --body-file 은 함께 쓸 수 없습니다.");
  }
  if (args.body !== undefined) return args.body;
  if (args.bodyFile !== undefined) return await readBodyFile(args.bodyFile);
  if (!process.stdin.isTTY) return await readStdin();
  throw new WikiApiError("본문은 --body, --body-file 또는 stdin 으로 입력해야 합니다.");
}

function parseIntegerOption(optionLabel: string, value: string | undefined): number {
  if (value === undefined) throw new CliParseError(`${optionLabel} 값이 필요합니다.`);
  const parsedValue = Number(value);
  if (!Number.isInteger(parsedValue)) throw new CliParseError(`${optionLabel} 값은 정수여야 합니다.`);
  return parsedValue;
}

function takeValue(argv: string[], index: number, optionLabel: string): [string, number] {
  const value = argv[index + 1];
  if (value === undefined) throw new CliParseError(`${optionLabel} 값이 필요합니다.`);
  return [value, index + 2];
}

function ensureNoExtra(argv: string[], index: number): void {
  if (index < argv.length) throw new CliParseError(`알 수 없는 인자: ${argv[index]}`);
}

function parseArgv(argv: string[]): CliArgs {
  let index = 0;
  let noBrowser = false;
  while (argv[index] === "--no-browser") {
    noBrowser = true;
    index += 1;
  }

  const command = argv[index];
  if (command === undefined) throw new CliParseError("명령이 필요합니다.");
  index += 1;

  if (command === "read" || command === "children" || command === "delete") {
    const topic = argv[index];
    if (topic === undefined) throw new CliParseError(`${command} 명령에는 topic 이 필요합니다.`);
    index += 1;

    if (command === "delete") {
      let baseVersion: number | undefined;
      while (index < argv.length) {
        const optionLabel = argv[index];
        if (optionLabel === "--base-version") {
          baseVersion = parseIntegerOption(optionLabel, argv[index + 1]);
          index += 2;
        } else {
          throw new CliParseError(`알 수 없는 인자: ${optionLabel}`);
        }
      }
      return baseVersion === undefined ? { command, noBrowser, topic } : { command, noBrowser, topic, baseVersion };
    }

    ensureNoExtra(argv, index);
    return { command, noBrowser, topic };
  }

  if (command === "search") {
    const keyword = argv[index];
    if (keyword === undefined) throw new CliParseError("search 명령에는 keyword 가 필요합니다.");
    index += 1;
    ensureNoExtra(argv, index);
    return { command, noBrowser, keyword };
  }

  if (command === "toc" || command === "rootmap" || command === "lint") {
    ensureNoExtra(argv, index);
    return { command, noBrowser };
  }

  if (command === "write") {
    const topic = argv[index];
    if (topic === undefined) throw new CliParseError("write 명령에는 topic 이 필요합니다.");
    index += 1;

    let title: string | undefined;
    let summary: string | undefined;
    let body: string | undefined;
    let bodyFile: string | undefined;
    let baseVersion: number | undefined;
    let parent: string | undefined;

    while (index < argv.length) {
      const optionLabel = argv[index];
      if (optionLabel === "--title") {
        [title, index] = takeValue(argv, index, optionLabel);
      } else if (optionLabel === "--summary") {
        [summary, index] = takeValue(argv, index, optionLabel);
      } else if (optionLabel === "--body") {
        [body, index] = takeValue(argv, index, optionLabel);
      } else if (optionLabel === "--body-file") {
        [bodyFile, index] = takeValue(argv, index, optionLabel);
      } else if (optionLabel === "--base-version") {
        baseVersion = parseIntegerOption(optionLabel, argv[index + 1]);
        index += 2;
      } else if (optionLabel === "--parent") {
        [parent, index] = takeValue(argv, index, optionLabel);
      } else {
        throw new CliParseError(`알 수 없는 인자: ${optionLabel}`);
      }
    }

    if (title === undefined) throw new CliParseError("write 명령에는 --title 이 필요합니다.");
    if (summary === undefined) throw new CliParseError("write 명령에는 --summary 가 필요합니다.");

    return { command, noBrowser, topic, title, summary, body, bodyFile, baseVersion, parent };
  }

  if (command === "move") {
    const topic = argv[index];
    if (topic === undefined) throw new CliParseError("move 명령에는 topic 이 필요합니다.");
    index += 1;

    let parent: string | undefined;
    let root = false;
    while (index < argv.length) {
      const optionLabel = argv[index];
      if (optionLabel === "--parent") {
        [parent, index] = takeValue(argv, index, optionLabel);
      } else if (optionLabel === "--root") {
        root = true;
        index += 1;
      } else {
        throw new CliParseError(`알 수 없는 인자: ${optionLabel}`);
      }
    }

    if ((parent !== undefined && root) || (parent === undefined && !root)) {
      throw new CliParseError("move 명령에는 --parent 또는 --root 중 하나가 필요합니다.");
    }
    return { command, noBrowser, topic, parent, root };
  }

  throw new CliParseError(`알 수 없는 명령: ${command}`);
}

function asTopicArgs(args: CliArgs): TopicArgs {
  return args as TopicArgs;
}

function asSearchArgs(args: CliArgs): SearchArgs {
  return args as SearchArgs;
}

function asWriteArgs(args: CliArgs): WriteArgs {
  return args as WriteArgs;
}

function asMoveArgs(args: CliArgs): MoveArgs {
  return args as MoveArgs;
}

async function runCommand(args: CliArgs, token: string): Promise<unknown> {
  if (args.command === "read") return await callService("read", [asTopicArgs(args).topic], token);
  if (args.command === "search") return await callService("search", [asSearchArgs(args).keyword], token);
  if (args.command === "toc") return await callService("toc", [], token);
  if (args.command === "rootmap") return await callService("rootMap", [], token);
  if (args.command === "children") return await callService("children", [asTopicArgs(args).topic], token);
  if (args.command === "write") {
    const writeArgs = asWriteArgs(args);
    const inputData: Record<string, unknown> = {
      topic: writeArgs.topic,
      title: writeArgs.title,
      summary: writeArgs.summary,
      body: await readBodyArg(writeArgs),
    };
    if (writeArgs.baseVersion !== undefined) inputData["baseVersion"] = writeArgs.baseVersion;
    if (writeArgs.parent !== undefined) inputData["parentTopic"] = writeArgs.parent;
    return await writePage(inputData, token);
  }
  if (args.command === "delete") {
    const topicArgs = asTopicArgs(args);
    const deleteInput: Record<string, unknown> = { topic: topicArgs.topic };
    const baseVersion = (args as { readonly baseVersion?: number }).baseVersion;
    if (baseVersion !== undefined) deleteInput["baseVersion"] = baseVersion;
    return await callService("delete", [deleteInput], token);
  }
  if (args.command === "move") {
    const moveArgs = asMoveArgs(args);
    const parentTopic = moveArgs.root ? null : moveArgs.parent;
    return await callService("move", [{ topic: moveArgs.topic, parentTopic }], token);
  }
  if (args.command === "lint") return await callService("lint", [], token);
  throw new WikiApiError(`알 수 없는 명령: ${args.command}`);
}

async function main(argv: string[]): Promise<number> {
  let args: CliArgs;
  try {
    args = parseArgv(argv);
  } catch (error) {
    if (error instanceof CliParseError) {
      console.error(error.message);
      return 2;
    }
    throw error;
  }

  const allowBrowser = !args.noBrowser;

  try {
    let token = await getToken(allowBrowser);
    if (token === null) return 1;
    let result: unknown;
    try {
      result = await runCommand(args, token);
    } catch (error) {
      if (!(error instanceof WikiAuthExpired)) throw error;
      if (!allowBrowser) throw error;
      token = await browserLogin();
      result = await runCommand(args, token);
    }
    printJson(result);
    return 0;
  } catch (error) {
    if (error instanceof WikiWriteConflict) {
      // 충돌은 실패(비0 종료)로 알리되, 재통합에 필요한 최신 본문을 stdout 으로 함께 전달.
      printJson({ conflict: true, message: error.message, latest: error.latest });
      return 3;
    }
    if (error instanceof WikiAuthError) {
      console.error(`위키 인증 오류: ${error.message}`);
      return 2;
    }
    if (error instanceof WikiApiError) {
      console.error(`위키 API 오류: ${error.message}`);
      return 2;
    }
    throw error;
  }
}

if (import.meta.main) {
  process.exit(await main(process.argv.slice(2)));
}
